src/rocksdb/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+                        metrics.write_latency_us.observe(f64::cast_lossy(latency));
+                        metrics
+                            .write_batch_size
+                            .observe(f64::cast_lossy(batch_size));
+                    }
+                    Err(e) => {
+                        let _ = response_sender.send(Err(Error::RocksDB(e)));
+                        return;
+                    }
+                }
+                // debug!("finished writing batch size({size}) for worker {worker_id}, source {source_id}");
+
+                let _ = response_sender.send(Ok(result));
+            }
+
+            // Gracefully cleanup if the `RocksDBInstance` has gone away.
+            db.cancel_all_background_work(true);
+            drop(db);
+        });
+
+        Ok(Self { tx })
+    }
+
+    /// Upsert a batch of values, returning the keys and values back, along
+    /// with previous values. If this errors, the `RocksDB` instance becomes
+    /// unusable.
+    pub async fn upsert(
+        &mut self,
+        batch: Vec<UpsertValue<K, V>>,
+    ) -> Result<Vec<UpsertResult<K, V>>, Error> {
+        if batch.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let (tx, rx) = oneshot_channel();
+        self.tx
+            .send(Command::Upsert {
+                batch,
+                response_sender: tx,
+            })
+            .await
+            .map_err(|_| Error::RocksDBThreadGoneAway)?;
+
+        // We also unwrap all rocksdb errors here.
+        rx.await.map_err(|_| Error::RocksDBThreadGoneAway)?
+    }
+
+    /// Gracefully shutdown RocksDB. Can error if the instance
+    /// is already shut down or errored.
+    pub async fn close(self) -> Result<(), Error> {
+        let (tx, rx) = oneshot_channel();
+        self.tx
+            .send(Command::Shutdown { done_sender: tx })
+            .await
+            .map_err(|_| Error::RocksDBThreadGoneAway)?;
+
+        let _ = rx.await;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mz_ore::metrics::HistogramVecExt;
+    use prometheus::{HistogramOpts, HistogramVec};
+
+    fn metrics_for_tests() -> Result<RocksDBMetrics, anyhow::Error> {
+        let fake_hist_vec =
+            HistogramVec::new(HistogramOpts::new("fake", "fake_help"), &["fake_label"])?;
+
+        Ok(RocksDBMetrics {
+            multi_get_latency_us: fake_hist_vec
+                .get_delete_on_drop_histogram(vec!["one".to_string()]),
+            multi_get_batch_size: fake_hist_vec
+                .get_delete_on_drop_histogram(vec!["two".to_string()]),
+            write_latency_us: fake_hist_vec.get_delete_on_drop_histogram(vec!["three".to_string()]),
+            write_batch_size: fake_hist_vec.get_delete_on_drop_histogram(vec!["four".to_string()]),
+        })
+    }
+
+    #[tokio::test]
+    #[cfg_attr(miri, ignore)] // unsupported operation: can't call foreign function `rocksdb_create_default_env` on OS `linux`
```

Comment:
If RocksDB is just wholly incompatible with miri, let's just exclude this entire crate from the miri runs (ci/test/cargo-test-miri.sh).

Reply:
Ah didn't know this was possible!